File: backend/src/main/java/kr/co/programmers/collabond/core/auth/oauth2/OAuth2UserService.java
package kr.co.programmers.collabond.core.auth.oauth2;

import jakarta.transaction.Transactional;
import kr.co.programmers.collabond.api.user.domain.Role;
import kr.co.programmers.collabond.api.user.domain.User;
import kr.co.programmers.collabond.api.user.infrastructure.UserRepository;
import kr.co.programmers.collabond.shared.exception.ErrorCode;
import kr.co.programmers.collabond.shared.exception.custom.DuplicatedException;
import kr.co.programmers.collabond.shared.exception.InvalidException;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.security.oauth2.client.userinfo.DefaultOAuth2UserService;
import org.springframework.security.oauth2.client.userinfo.OAuth2UserRequest;
import org.springframework.security.oauth2.core.user.OAuth2User;
import org.springframework.stereotype.Service;

import java.util.Optional;

@Slf4j
@Service
@RequiredArgsConstructor
public class OAuth2UserService extends DefaultOAuth2UserService {

    private final UserRepository userRepository;

    @Transactional
    @Override
    public OAuth2User loadUser(OAuth2UserRequest userRequest) {
        // 기본 정보 가져오기
        OAuth2User oAuth2User = super.loadUser(userRequest);

        String provider = userRequest.getClientRegistration().getRegistrationId();

        OAuth2ResponseDto oAuth2ResponseDto = null;

        switch (provider) {
            case "google":
                oAuth2ResponseDto = new GoogleResponseDto(oAuth2User.getAttributes());
                break;
            case "naver":
                oAuth2ResponseDto = new NaverResponseDto(oAuth2User.getAttributes());
                break;
            case "kakao":
                oAuth2ResponseDto = new KakaoResponseDto(oAuth2User.getAttributes());
                break;
            default:
                throw new InvalidException(ErrorCode.INVALID_PROVIDER);
        }

        // providerId로 사용자 찾기
        String providerId = oAuth2ResponseDto.getProvider() + "_" + oAuth2ResponseDto.getProviderId();
        Optional<User> userByProviderId = userRepository.findByProviderId(providerId);
        if (userByProviderId.isPresent()) {
            return OAuth2Mapper.toOAuth2UserInfo(userByProviderId.get());
        }

        String email = oAuth2ResponseDto.getEmail();
        userRepository.findByEmail(email).ifPresent(user -> {
            throw new DuplicatedException(ErrorCode.DUPLICATED_USER_EMAIL);
        });

        User newUser = userRepository.save(OAuth2Mapper.toUser(oAuth2ResponseDto, Role.ROLE_TMP));

        return OAuth2Mapper.toOAuth2UserInfo(newUser);
    }

}

File: backend/src/main/java/kr/co/programmers/collabond/core/auth/oauth2/OAuth2UserService.kt
package kr.co.programmers.collabond.core.auth.oauth2

import jakarta.transaction.Transactional
import kr.co.programmers.collabond.api.user.domain.Role
import kr.co.programmers.collabond.api.user.infrastructure.UserRepository
import kr.co.programmers.collabond.core.auth.oauth2.OAuth2Mapper.toOAuth2UserInfo
import kr.co.programmers.collabond.core.auth.oauth2.OAuth2Mapper.toUser
import kr.co.programmers.collabond.shared.exception.ErrorCode
import kr.co.programmers.collabond.shared.exception.DuplicatedException
import kr.co.programmers.collabond.shared.exception.InvalidException
import org.slf4j.LoggerFactory
import org.springframework.security.oauth2.client.userinfo.DefaultOAuth2UserService
import org.springframework.security.oauth2.client.userinfo.OAuth2UserRequest
import org.springframework.security.oauth2.core.user.OAuth2User
import org.springframework.stereotype.Service

@Service
class OAuth2UserService(
    private val userRepository: UserRepository
) : DefaultOAuth2UserService() {

    private val logger = LoggerFactory.getLogger(OAuth2UserService::class.java)

    @Transactional
    override fun loadUser(userRequest: OAuth2UserRequest): OAuth2User {
        // 기본 정보 가져오기
        val oAuth2User = super.loadUser(userRequest)
        val provider = userRequest.clientRegistration.registrationId

        val oAuth2ResponseDto = when (provider) {
            "google" -> GoogleResponseDto(oAuth2User.attributes)
            "naver" -> NaverResponseDto(oAuth2User.attributes)
            "kakao" -> KakaoResponseDto(oAuth2User.attributes)
            else -> throw InvalidException(ErrorCode.INVALID_PROVIDER)
        }

        // providerId로 사용자 찾기
        val providerId = "${oAuth2ResponseDto.provider}_${oAuth2ResponseDto.providerId}"

        userRepository.findByProviderId(providerId)?.let { user ->
            return toOAuth2UserInfo(user)
        }

        // 이메일 중복 체크
        userRepository.findByEmail(oAuth2ResponseDto.email)?.let {
            throw DuplicatedException(ErrorCode.DUPLICATED_USER_EMAIL)
        }

        val newUser = userRepository.save(
            toUser(oAuth2ResponseDto, Role.ROLE_TMP)
        )

        return toOAuth2UserInfo(newUser)
    }
}